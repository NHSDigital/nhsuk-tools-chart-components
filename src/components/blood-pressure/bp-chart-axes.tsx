import {
  BpClassificationLevelType,
  BpClassificationLevel,
} from '@custom-types/blood-pressure.types';
import { BpBounds } from '@custom-types/classifications.types';
import { UpperBounds } from '@custom-types/threshold.types';
import React from 'react';
import './styles/bp-chart-axes.module.scss';

export enum AxisType {
  Systolic,
  Diastolic,
}

type BpGraphAxisLabelProps = Readonly<{
  axisType: AxisType;
  labelText: string;
  offsetPercent: number;
  applyLabelSpecificStyles?: boolean;
}>;

type BpGraphAxisProps = Readonly<{
  axisType: AxisType;
  upperBounds: UpperBounds<string>;
  lowerBound: number;
}>;

type BpGraphAxesProps = Readonly<{
  bounds: BpBounds;
}>;

export function BpGraphAxisValueLabel({
  axisType,
  labelText,
  offsetPercent,
  applyLabelSpecificStyles = true,
}: BpGraphAxisLabelProps): JSX.Element {
  const classPrefix = `${AxisType[axisType].toLowerCase()}-axis`;
  const genericMarkerClass = `nhsuk-bp-${classPrefix}-marker`;
  const specificMarkerClass = `nhsuk-bp-${classPrefix}-marker-${labelText}`;

  return (
    <div
      className={
        applyLabelSpecificStyles
          ? `${genericMarkerClass} ${specificMarkerClass}`
          : `${genericMarkerClass}`
      }
      style={
        axisType === AxisType.Systolic
          ? { bottom: `${offsetPercent}%` }
          : { left: `${offsetPercent}%` }
      }>
      <div className={`nhsuk-bp-${classPrefix}-label`}>{labelText}</div>
    </div>
  );
}

function boundaryPercentage(
  classificationUpperBoundValue: number,
  topUpperBoundValue: number,
  lowerBound: number
): number {
  return (
    ((classificationUpperBoundValue - lowerBound) /
      (topUpperBoundValue - lowerBound)) *
    100
  );
}

function classificationBoundaryPercentage(
  classification: BpClassificationLevelType,
  upperBounds: UpperBounds<string>,
  lowerBound: number
): number {
  return boundaryPercentage(
    upperBounds[classification].value!,
    upperBounds[BpClassificationLevel.High].value!,
    lowerBound
  );
}

function BpGraphAxis({
  axisType,
  upperBounds,
  lowerBound,
}: BpGraphAxisProps): JSX.Element {
  const extraDiastolicLabelConfig = {
    slightlyRaisedValueToCheck: 85,
    slightlyRaisedValueToAdd: 90,
  };
  return (
    <div>
      <BpGraphAxisValueLabel
        labelText={lowerBound.toString()}
        offsetPercent={0}
        axisType={axisType}
      />
      <BpGraphAxisValueLabel
        labelText={upperBounds[BpClassificationLevel.Low].value!.toString()}
        offsetPercent={classificationBoundaryPercentage(
          BpClassificationLevel.Low,
          upperBounds,
          lowerBound
        )}
        axisType={axisType}
      />
      <BpGraphAxisValueLabel
        labelText={upperBounds[BpClassificationLevel.Healthy].value!.toString()}
        offsetPercent={classificationBoundaryPercentage(
          BpClassificationLevel.Healthy,
          upperBounds,
          lowerBound
        )}
        axisType={axisType}
      />
      <BpGraphAxisValueLabel
        labelText={upperBounds[
          BpClassificationLevel.SlightlyRaised
        ].value!.toString()}
        offsetPercent={classificationBoundaryPercentage(
          BpClassificationLevel.SlightlyRaised,
          upperBounds,
          lowerBound
        )}
        axisType={axisType}
        applyLabelSpecificStyles={
          upperBounds[BpClassificationLevel.SlightlyRaised].value ===
          extraDiastolicLabelConfig.slightlyRaisedValueToCheck
        }
      />
      {axisType === AxisType.Diastolic &&
        upperBounds[BpClassificationLevel.SlightlyRaised].value ===
          extraDiastolicLabelConfig.slightlyRaisedValueToCheck && (
          <BpGraphAxisValueLabel
            labelText={extraDiastolicLabelConfig.slightlyRaisedValueToAdd.toString()}
            offsetPercent={boundaryPercentage(
              extraDiastolicLabelConfig.slightlyRaisedValueToAdd,
              upperBounds[BpClassificationLevel.High].value!,
              lowerBound
            )}
            axisType={axisType}
          />
        )}
      <BpGraphAxisValueLabel
        labelText={upperBounds[BpClassificationLevel.High].value!.toString()}
        offsetPercent={100}
        axisType={axisType}
      />
    </div>
  );
}

export function BpGraphAxes({ bounds }: BpGraphAxesProps): JSX.Element {
  return (
    <>
      <BpGraphAxis
        axisType={AxisType.Systolic}
        upperBounds={bounds.systolic.upper}
        lowerBound={bounds.systolic.lower}
      />
      <BpGraphAxis
        axisType={AxisType.Diastolic}
        upperBounds={bounds.diastolic.upper}
        lowerBound={bounds.diastolic.lower}
      />
    </>
  );
}
