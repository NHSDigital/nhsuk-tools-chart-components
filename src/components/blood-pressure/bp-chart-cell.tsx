import { BpCellConfiguration } from '@custom-types/blood-pressure.types';
import React from 'react';

import './styles/bp-chart-cell.module.scss';

export function BpChartCell({
  firstInRow = false,
  firstInColumn = false,
}: Readonly<BpCellConfiguration>): JSX.Element {
  let classNames = 'nhsuk-bp-graph-cell';

  if (firstInRow) {
    classNames += ` ${'nhsuk-bp-first-in-row'}`;
  }

  if (firstInColumn) {
    classNames += ` ${'nhsuk-bp-first-in-column'}`;
  }

  return <div className={classNames} />;
}
