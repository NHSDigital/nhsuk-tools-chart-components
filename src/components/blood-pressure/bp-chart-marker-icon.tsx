import React, { JSX } from 'react';

export function BpGraphMarkerIcon(): JSX.Element {
  return (
    <svg viewBox="0 0 32 32" fill="none" xmlns="http://www.w3.org/2000/svg">
      <g clipPath="url(#clip0_1346_102759)">
        <path
          d="M5.18718 8.36931L4.48008 9.07641L5.18718 9.78352L11.3744 15.9707L5.18718 22.1579L4.48008 22.865L5.18718 23.5721L8.36916 26.7541L9.07627 27.4612L9.78338 26.7541L15.9706 20.5669L22.1577 26.7541L22.8649 27.4612L23.572 26.7541L26.7539 23.5721L27.461 22.865L26.7539 22.1579L20.5668 15.9707L26.7539 9.78352L27.461 9.07641L26.7539 8.36931L23.572 5.18732L22.8649 4.48022L22.1577 5.18732L15.9706 11.3745L9.78338 5.18732L9.07627 4.48022L8.36916 5.18732L5.18718 8.36931Z"
          fill="black"
          stroke="white"
          strokeWidth="2"
        />
      </g>
    </svg>
  );
}
