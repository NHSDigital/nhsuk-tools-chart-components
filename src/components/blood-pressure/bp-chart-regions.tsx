import {
  BpClassificationLevel,
  BpClassificationLevelType,
} from '@custom-types/blood-pressure.types';
import { BpBounds } from '@custom-types/classifications.types';
import { UpperBounds } from '@custom-types/threshold.types';
import React from 'react';
import './styles/bp-chart-region-boundaries.module.scss';

export type BpGraphBoundaryProps = Readonly<{
  bounds: BpBounds;
  renderAsBoundary?: boolean;
}>;

function getClassificationBoundaryPercentage(
  lowerBound: number,
  upperBounds: UpperBounds<string>,
  classification: BpClassificationLevelType
): number {
  const normalisedBound = upperBounds[classification].value! - lowerBound;
  const normalisedUpperBound = upperBounds[3].value! - lowerBound;

  return (normalisedBound / normalisedUpperBound) * 100;
}

function getStyleForClassification(
  classification: BpClassificationLevelType
): string {
  return `nhsuk-bp-classification-${classification.toLowerCase()}`;
}

export function RegionsForClassification({
  classification,
  bounds,
  renderAsBoundary = false,
}: Readonly<
  BpGraphBoundaryProps & {
    classification: BpClassificationLevelType;
  }
>): JSX.Element {
  const width = getClassificationBoundaryPercentage(
    bounds.diastolic.lower,
    bounds.diastolic.upper,
    classification
  );
  const height = getClassificationBoundaryPercentage(
    bounds.systolic.lower,
    bounds.systolic.upper,
    classification
  );

  return (
    <div
      className={
        renderAsBoundary
          ? 'nhsuk-bp-region-boundary'
          : getStyleForClassification(classification)
      }
      style={{
        width: `${width}%`,
        height: `${height}%`,
      }}
    />
  );
}

export function BpGraphRegions({
  bounds,
  renderAsBoundary = false,
}: BpGraphBoundaryProps): JSX.Element {
  return (
    <>
      <RegionsForClassification
        classification={BpClassificationLevel.Low}
        {...{ bounds, renderAsBoundary }}
      />
      <RegionsForClassification
        classification={BpClassificationLevel.Healthy}
        {...{ bounds, renderAsBoundary }}
      />
      <RegionsForClassification
        classification={BpClassificationLevel.SlightlyRaised}
        {...{ bounds, renderAsBoundary }}
      />
      {!renderAsBoundary && (
        <RegionsForClassification
          classification={BpClassificationLevel.High}
          {...{ bounds, renderAsBoundary }}
        />
      )}
    </>
  );
}
