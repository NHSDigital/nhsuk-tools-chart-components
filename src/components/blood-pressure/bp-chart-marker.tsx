import Big from 'big.js';
import React, { JSX } from 'react';
import { BpGraphMarkerIcon } from './bp-chart-marker-icon';
import './styles/bp-chart-marker.module.scss';

// Changes rounding strategy to banker's rounding for unbias rounding
Big.RM = Big.roundHalfEven;

type BpGraphMarkerProps = Readonly<{
  systolicReading: number;
  diastolicReading: number;
  systolicLowerBound: number;
  systolicUpperBound: number;
  diastolicLowerBound: number;
  diastolicUpperBound: number;
}>;

export function BpGraphMarker({
  systolicReading,
  diastolicReading,
  systolicLowerBound,
  systolicUpperBound,
  diastolicLowerBound,
  diastolicUpperBound,
}: BpGraphMarkerProps): JSX.Element {
  const systolicDifference = systolicUpperBound - systolicLowerBound;
  const diastolicDifference = diastolicUpperBound - diastolicLowerBound;

  const bottomPercent = new Big(
    (systolicReading - systolicLowerBound) / systolicDifference
  ).times(100);
  const leftPercent = new Big(
    (diastolicReading - diastolicLowerBound) / diastolicDifference
  ).times(100);

  const roundingPrecision = 1;

  const roundedBottomPercent = bottomPercent.round(roundingPrecision);
  const roundedLeftPercent = leftPercent.round(roundingPrecision);

  return (
    <div
      className="nhsuk-bp-graph-marker"
      style={{
        left: `${roundedLeftPercent}%`,
        bottom: `${roundedBottomPercent}%`,
      }}>
      <BpGraphMarkerIcon />
    </div>
  );
}
