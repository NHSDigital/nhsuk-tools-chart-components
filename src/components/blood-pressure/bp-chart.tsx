import { BpBounds } from '@custom-types/classifications.types';
import {
  BpGraphLayout,
  BpReadingValue,
} from '@custom-types/blood-pressure.types';
import React from 'react';
import { BpGraphAxes } from './bp-chart-axes';
import { BpGraphMarker } from './bp-chart-marker';
import { BpGraphRegions } from './bp-chart-regions';
import { BpChartCell } from './bp-chart-cell';
import './styles/bp-chart.module.scss';

/**
 * The properties for @BpChart component
 *
 * @param {BpBounds} bounds - The blood pressure classification ranges
 * @param {BpGraphLayout} graphLayout - The graph layout contains rows configuration and systolic/diastolic chart labels
 * @param {BpReadingValue} readings - The systolic and diastolic blood pressure readings
 */
export type BpChartProps = {
  bounds: BpBounds;
  graphLayout: BpGraphLayout;
  reading: BpReadingValue;
};

/**
 * A component to display the blood pressure chart
 *
 * @typedef {object} BpChartProps
 * @property {BpBounds} bounds - The blood pressure classification ranges
 * @property {BpGraphLayout} graphLayout - The graph layout contains rows configuration and systolic/diastolic chart labels
 * @property {BpReadingValue} readings - The systolic and diastolic blood pressure readings
 *
 * @param {BpChartProps} props
 * @returns {JSX.Element}
 */
export function BpChart({
  bounds,
  graphLayout,
  reading,
}: Readonly<BpChartProps>): JSX.Element {
  const { systolic, diastolic } = reading;
  const { systolicLabel, diastolicLabel } = graphLayout;

  return (
    <div className="nhsuk-body-s">
      <div className="nhsuk-bp-systolic-legend-wrapper" aria-hidden="true">
        <div className="nhsuk-bp-systolic-legend-marker" />
        <span>{systolicLabel}</span>
      </div>

      <div className="nhsuk-bp-graph" aria-hidden="true">
        <BpGraphAxes bounds={bounds} />
        <BpGraphMarker
          systolicReading={systolic}
          diastolicReading={diastolic}
          systolicLowerBound={bounds.systolic.lower}
          systolicUpperBound={bounds.systolic.upper.high.value!}
          diastolicLowerBound={bounds.diastolic.lower}
          diastolicUpperBound={bounds.diastolic.upper.high.value!}
        />
        <BpGraphRegions bounds={bounds} />
        {graphLayout.rows.map((rowConfig) => (
          <div key={rowConfig.id} className="nhsuk-bp-graph-row">
            {rowConfig.cells.map((cellProps) => (
              <BpChartCell key={cellProps.id} {...cellProps} />
            ))}
          </div>
        ))}
        <BpGraphRegions bounds={bounds} renderAsBoundary />
      </div>

      <div className="nhsuk-bp-diastolic-legend-wrapper" aria-hidden="true">
        <div className="nhsuk-bp-diastolic-legend-marker" />
        <span>{diastolicLabel}</span>
      </div>
    </div>
  );
}
