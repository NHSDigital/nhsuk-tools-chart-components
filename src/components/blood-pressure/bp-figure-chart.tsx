import React, { JSX } from 'react';
import { ChartLegend, ChartLegendKey } from '@components/chart-legend';
import { BpBounds } from '@custom-types/classifications.types';
import {
  BpGraphLayout,
  BpLegendKeys,
  BpReadingValue,
} from '@custom-types/blood-pressure.types';
import { BpChart } from './bp-chart';
import { BpGraphMarkerIcon } from './bp-chart-marker-icon';
import './styles/bp-chart-panel.module.scss';

const populateBpChartLegendItems = (
  legendKeys: BpLegendKeys,
  legendMarkerText: string
): ChartLegendKey[] =>
  Object.entries(legendKeys)
    .map(
      ([classification, label]): ChartLegendKey => ({
        text: label,
        className: `nhsuk-bp-legend-key-classification-${classification}`,
      })
    )
    .concat({
      text: legendMarkerText,
      className: 'nhsuk-bp-legend-key-marker',
      element: <BpGraphMarkerIcon />,
    });

/**
 * The properties for @BpChartFigure component
 *
 * @param {BpBounds} bounds - The blood pressure classification ranges
 * @param {BpGraphLayout} graphLayout - The graph layout with row configuration and systolic/diastolic chart labels
 * @param {BpReadingValue} readings - The systolic and diastolic blood pressure readings
 * @param {string }ariaLabel - The aria label for the chart figure
 * @param {BpLegendKeys} legendKeys - The legend keys for the chart figure
 * @param {string } legendMarkerText - The legend marker text
 */
export type BpChartFigureProps = {
  bounds: BpBounds;
  graphLayout: BpGraphLayout;
  reading: BpReadingValue;
  ariaLabel: string;
  legendKeys: BpLegendKeys;
  legendMarkerText: string;
};

/**
 * A component to display the blood pressure chart figure
 *
 * @typedef {object} BpChartFigureProps
 * @property {BpBounds} bounds - The blood pressure classification ranges
 * @property {BpGraphLayout} graphLayout - The graph layout with row configuration and systolic/diastolic chart labels
 * @property {BpReadingValue} readings - The systolic and diastolic blood pressure readings
 * @property {string }ariaLabel - The aria label for the chart figure
 * @property {BpLegendKeys} legendKeys - The legend keys for the chart figure
 * @property {string } legendMarkerText - The legend marker text
 *
 * @param {BpChartFigureProps} props
 * @returns {JSX.Element}
 */
export function BpChartFigure({
  bounds,
  graphLayout,
  reading,
  ariaLabel,
  legendKeys,
  legendMarkerText,
}: Readonly<BpChartFigureProps>): JSX.Element {
  const bpLegendKeys = populateBpChartLegendItems(legendKeys, legendMarkerText);
  return (
    <figure className="nhsuk-u-margin-0">
      <figcaption className="nhsuk-u-visually-hidden">{ariaLabel}</figcaption>
      <BpChart bounds={bounds} reading={reading} graphLayout={graphLayout} />

      <ChartLegend legendKeys={bpLegendKeys} />
    </figure>
  );
}
