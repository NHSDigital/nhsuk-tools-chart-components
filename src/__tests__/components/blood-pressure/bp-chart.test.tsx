import { BpChart } from '@components/blood-pressure/bp-chart';
import { render } from '@testing-library/react';
import '@testing-library/jest-dom';
import {
  BpGraphLayout,
  BpReadingValue,
} from '@custom-types/blood-pressure.types';
import { BpBounds } from '@custom-types/classifications.types';
import {
  mockedBpClassificationBounds as mockedClassificationBounds,
  mockedBpGraphLayout,
  mockedBptReading,
} from '@tests/__helpers__/mocks/bp-results';
import React from 'react';

const systolicAxis = [
  { value: 70, offset: 0 },
  { value: 90, offset: 18.181818181818183 },
  { value: 120, offset: 45.45454545454545 },
  { value: 140, offset: 63.63636363636363 },
  { value: 180, offset: 100 },
];

const diastolicAxis = [
  { value: 40, offset: 0 },
  { value: 60, offset: 25 },
  { value: 80, offset: 50 },
  { value: 90, offset: 62.5 },
  { value: 120, offset: 100 },
];

const bptClassificationRegions = [
  { name: 'classification-0', width: 25, height: 18.181818181818183 },
  { name: 'classification-1', width: 50, height: 45.45454545454545 },
  { name: 'classification-2', width: 62.5, height: 63.63636363636363 },
  { name: 'classification-3', width: 100, height: 100 },
];

const btpRegionBoundaries = [
  { width: 25, height: 18.181818181818183 },
  { width: 50, height: 45.45454545454545 },
  { width: 62.5, height: 63.63636363636363 },
];

function ResultBpChartTestApp({
  classificationBounds = mockedClassificationBounds,
  reading = mockedBptReading,
  layout = mockedBpGraphLayout,
}: Readonly<{
  classificationBounds?: BpBounds;
  reading?: BpReadingValue;
  layout?: BpGraphLayout;
}>): JSX.Element {
  return (
    <BpChart
      bounds={classificationBounds}
      reading={reading}
      graphLayout={layout}
    />
  );
}

describe('The Blood pressure result chart component', () => {
  it('Renders with the expected content', () => {
    const { container } = render(<ResultBpChartTestApp />);
    expect(container).toMatchSnapshot();
  });

  it('Renders the systolic chart axis ', () => {
    const { container } = render(<ResultBpChartTestApp />);

    const axisElements = container?.querySelectorAll(
      '.nhsuk-bp-systolic-axis-marker'
    ) as NodeListOf<HTMLElement>;

    expect(axisElements.length).toBe(systolicAxis.length);

    axisElements.forEach((sysAxis, index) => {
      expect(sysAxis.style.bottom).toBe(`${systolicAxis[index].offset}%`);

      const axisLabelEl = sysAxis.querySelector(
        '.nhsuk-bp-systolic-axis-label'
      );
      expect(axisLabelEl).not.toBeNull();

      expect(axisLabelEl?.innerHTML).toBe(systolicAxis[index].value.toString());
    });
  });

  it('Renders the systolic legend marker and label ', () => {
    const { container } = render(<ResultBpChartTestApp />);

    const axisLegendMarkerEl = container?.querySelector(
      '.nhsuk-bp-systolic-legend-wrapper'
    );
    expect(axisLegendMarkerEl).not.toBeNull();
    const legendMarkerEl = axisLegendMarkerEl!.querySelector(
      '.nhsuk-bp-systolic-legend-marker'
    );

    expect(legendMarkerEl).not.toBeNull();
    const spanElement = axisLegendMarkerEl!.querySelector('span');

    expect(spanElement?.innerHTML).toBe(mockedBpGraphLayout.systolicLabel);
  });

  it('Renders the diastolic legend marker and label ', () => {
    const { container } = render(<ResultBpChartTestApp />);

    const axisLegendMarkerEl = container?.querySelector(
      '.nhsuk-bp-diastolic-legend-wrapper'
    );
    expect(axisLegendMarkerEl).not.toBeNull();
    const legendMarkerEl = axisLegendMarkerEl!.querySelector(
      '.nhsuk-bp-diastolic-legend-marker'
    );

    expect(legendMarkerEl).not.toBeNull();
    const spanElement = axisLegendMarkerEl!.querySelector('span');

    expect(spanElement?.innerHTML).toBe(mockedBpGraphLayout.diastolicLabel);
  });

  it('Renders the diastolic chart axis ', () => {
    const { container } = render(<ResultBpChartTestApp />);

    const axisElements = container?.querySelectorAll(
      '.nhsuk-bp-diastolic-axis-marker'
    ) as NodeListOf<HTMLElement>;

    expect(axisElements.length).toBe(diastolicAxis.length);

    axisElements.forEach((sysAxis, index) => {
      expect(sysAxis.style.left).toBe(`${diastolicAxis[index].offset}%`);

      const axisLabelEl = sysAxis.querySelector(
        '.nhsuk-bp-diastolic-axis-label'
      );
      expect(axisLabelEl).not.toBeNull();

      expect(axisLabelEl?.innerHTML).toBe(
        diastolicAxis[index].value.toString()
      );
    });
  });

  it('Renders the chart classification regions', () => {
    const { container } = render(<ResultBpChartTestApp />);

    bptClassificationRegions.forEach((region) => {
      const regionEl = container?.querySelector(
        `.nhsuk-bp-${region.name}`
      ) as HTMLElement;

      expect(regionEl).not.toBeNull();

      expect(regionEl.style.width).toBe(`${region.width}%`);
      expect(regionEl.style.height).toBe(`${region.height}%`);
    });
  });

  it('Renders the chart region boundaries', () => {
    const { container } = render(<ResultBpChartTestApp />);

    const regionBoundariesEl = container?.querySelectorAll(
      '.nhsuk-bp-region-boundary'
    ) as NodeListOf<HTMLElement>;

    expect(regionBoundariesEl.length).toBe(regionBoundariesEl.length);

    regionBoundariesEl.forEach((regionBoundary, index) => {
      expect(regionBoundary.style.width).toBe(
        `${btpRegionBoundaries[index].width}%`
      );
      expect(regionBoundary.style.height).toBe(
        `${btpRegionBoundaries[index].height}%`
      );
    });
  });

  it('Renders the chart layout', () => {
    const { container } = render(<ResultBpChartTestApp />);

    const rowsEl = container?.querySelectorAll('.nhsuk-bp-graph-row');
    expect(rowsEl).not.toBeNull();
    expect(rowsEl.length).toBe(mockedBpGraphLayout.rows.length);

    rowsEl.forEach((rowEl, rowIdx) => {
      const row = mockedBpGraphLayout.rows[rowIdx];

      const cellsEl = rowEl?.querySelectorAll('.nhsuk-bp-graph-cell');
      expect(cellsEl.length).toBe(row.cells.length);

      cellsEl.forEach((cellEl, cellIdx) => {
        const cell = row.cells[cellIdx];

        if (cell.firstInRow) {
          expect(cellEl.classList).toContain('nhsuk-bp-first-in-row');
        }

        if (cell.firstInColumn) {
          expect(cellEl.classList).toContain('nhsuk-bp-first-in-column');
        }
      });
    });
  });

  it.each`
    bpReading     | expectedOffset
    ${[100, 60]}  | ${[25, 27.3]}
    ${[110, 71]}  | ${[38.8, 36.4]}
    ${[120, 75]}  | ${[43.8, 45.5]}
    ${[124, 72]}  | ${[40, 49.1]}
    ${[145, 83]}  | ${[53.8, 68.2]}
    ${[160, 65]}  | ${[31.2, 81.8]}
    ${[180, 83]}  | ${[53.8, 100]}
    ${[200, 100]} | ${[75, 118.2]}
  `(
    'when reading value of $bpReading is provided, the chart should contain a chart marker positioned at $expectedOffset',
    async ({
      bpReading,
      expectedOffset,
    }: {
      bpReading: [number, number];
      expectedOffset: [number, number];
    }) => {
      const [systolic, diastolic] = bpReading;
      const { container } = render(
        <ResultBpChartTestApp
          reading={{
            systolic,
            diastolic,
          }}
        />
      );

      const [left, bottom] = expectedOffset;
      const markerEl = container?.querySelector(
        '.nhsuk-bp-graph-marker'
      ) as HTMLElement;

      expect(markerEl).not.toBeNull();

      expect(markerEl.style.left).toBe(`${left}%`);
      expect(markerEl.style.bottom).toBe(`${bottom}%`);
    }
  );
});
