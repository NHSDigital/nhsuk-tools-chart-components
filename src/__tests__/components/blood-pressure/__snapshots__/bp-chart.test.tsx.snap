// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Blood pressure result chart component Renders with the expected content 1`] = `
<div>
  <div
    class="nhsuk-body-s"
  >
    <div
      aria-hidden="true"
      class="nhsuk-bp-systolic-legend-wrapper"
    >
      <div
        class="nhsuk-bp-systolic-legend-marker"
      />
      <span>
        Systolic (high number)
      </span>
    </div>
    <div
      aria-hidden="true"
      class="nhsuk-bp-graph"
    >
      <div>
        <div
          class="nhsuk-bp-systolic-axis-marker nhsuk-bp-systolic-axis-marker-70"
          style="bottom: 0%;"
        >
          <div
            class="nhsuk-bp-systolic-axis-label"
          >
            70
          </div>
        </div>
        <div
          class="nhsuk-bp-systolic-axis-marker nhsuk-bp-systolic-axis-marker-90"
          style="bottom: 18.181818181818183%;"
        >
          <div
            class="nhsuk-bp-systolic-axis-label"
          >
            90
          </div>
        </div>
        <div
          class="nhsuk-bp-systolic-axis-marker nhsuk-bp-systolic-axis-marker-120"
          style="bottom: 45.45454545454545%;"
        >
          <div
            class="nhsuk-bp-systolic-axis-label"
          >
            120
          </div>
        </div>
        <div
          class="nhsuk-bp-systolic-axis-marker"
          style="bottom: 63.63636363636363%;"
        >
          <div
            class="nhsuk-bp-systolic-axis-label"
          >
            140
          </div>
        </div>
        <div
          class="nhsuk-bp-systolic-axis-marker nhsuk-bp-systolic-axis-marker-180"
          style="bottom: 100%;"
        >
          <div
            class="nhsuk-bp-systolic-axis-label"
          >
            180
          </div>
        </div>
      </div>
      <div>
        <div
          class="nhsuk-bp-diastolic-axis-marker nhsuk-bp-diastolic-axis-marker-40"
          style="left: 0%;"
        >
          <div
            class="nhsuk-bp-diastolic-axis-label"
          >
            40
          </div>
        </div>
        <div
          class="nhsuk-bp-diastolic-axis-marker nhsuk-bp-diastolic-axis-marker-60"
          style="left: 25%;"
        >
          <div
            class="nhsuk-bp-diastolic-axis-label"
          >
            60
          </div>
        </div>
        <div
          class="nhsuk-bp-diastolic-axis-marker nhsuk-bp-diastolic-axis-marker-80"
          style="left: 50%;"
        >
          <div
            class="nhsuk-bp-diastolic-axis-label"
          >
            80
          </div>
        </div>
        <div
          class="nhsuk-bp-diastolic-axis-marker"
          style="left: 62.5%;"
        >
          <div
            class="nhsuk-bp-diastolic-axis-label"
          >
            90
          </div>
        </div>
        <div
          class="nhsuk-bp-diastolic-axis-marker nhsuk-bp-diastolic-axis-marker-120"
          style="left: 100%;"
        >
          <div
            class="nhsuk-bp-diastolic-axis-label"
          >
            120
          </div>
        </div>
      </div>
      <div
        class="nhsuk-bp-graph-marker"
        style="left: 50%; bottom: 63.6%;"
      >
        <svg
          fill="none"
          viewBox="0 0 32 32"
          xmlns="http://www.w3.org/2000/svg"
        >
          <g
            clip-path="url(#clip0_1346_102759)"
          >
            <path
              d="M5.18718 8.36931L4.48008 9.07641L5.18718 9.78352L11.3744 15.9707L5.18718 22.1579L4.48008 22.865L5.18718 23.5721L8.36916 26.7541L9.07627 27.4612L9.78338 26.7541L15.9706 20.5669L22.1577 26.7541L22.8649 27.4612L23.572 26.7541L26.7539 23.5721L27.461 22.865L26.7539 22.1579L20.5668 15.9707L26.7539 9.78352L27.461 9.07641L26.7539 8.36931L23.572 5.18732L22.8649 4.48022L22.1577 5.18732L15.9706 11.3745L9.78338 5.18732L9.07627 4.48022L8.36916 5.18732L5.18718 8.36931Z"
              fill="black"
              stroke="white"
              stroke-width="2"
            />
          </g>
        </svg>
      </div>
      <div
        class="nhsuk-bp-classification-0"
        style="width: 25%; height: 18.181818181818183%;"
      />
      <div
        class="nhsuk-bp-classification-1"
        style="width: 50%; height: 45.45454545454545%;"
      />
      <div
        class="nhsuk-bp-classification-2"
        style="width: 62.5%; height: 63.63636363636363%;"
      />
      <div
        class="nhsuk-bp-classification-3"
        style="width: 100%; height: 100%;"
      />
      <div
        class="nhsuk-bp-graph-row"
      >
        <div
          class="nhsuk-bp-graph-cell nhsuk-bp-first-in-row nhsuk-bp-first-in-column"
        />
        <div
          class="nhsuk-bp-graph-cell nhsuk-bp-first-in-column"
        />
        <div
          class="nhsuk-bp-graph-cell nhsuk-bp-first-in-column"
        />
        <div
          class="nhsuk-bp-graph-cell nhsuk-bp-first-in-column"
        />
        <div
          class="nhsuk-bp-graph-cell nhsuk-bp-first-in-column"
        />
        <div
          class="nhsuk-bp-graph-cell nhsuk-bp-first-in-column"
        />
      </div>
      <div
        class="nhsuk-bp-region-boundary"
        style="width: 25%; height: 18.181818181818183%;"
      />
      <div
        class="nhsuk-bp-region-boundary"
        style="width: 50%; height: 45.45454545454545%;"
      />
      <div
        class="nhsuk-bp-region-boundary"
        style="width: 62.5%; height: 63.63636363636363%;"
      />
    </div>
    <div
      aria-hidden="true"
      class="nhsuk-bp-diastolic-legend-wrapper"
    >
      <div
        class="nhsuk-bp-diastolic-legend-marker"
      />
      <span>
        Diastolic (low number)
      </span>
    </div>
  </div>
</div>
`;
